File: src/desktop-edge-agent/src/FccDesktopAgent.Core/Adapter/Doms/Model/DomsFpMainState.cs
using FccDesktopAgent.Core.Adapter.Common;

namespace FccDesktopAgent.Core.Adapter.Doms.Model;

/// <summary>
/// DOMS FP_MAIN_STATE values. Maps raw integer codes from the DOMS JPL protocol
/// to canonical <see cref="PumpState"/> via <see cref="DomsFpMainStateExtensions.ToCanonicalPumpState"/>.
/// </summary>
public enum DomsFpMainState
{
    FpInoperative = 0,
    FpClosed = 1,
    FpIdle = 2,
    FpCalling = 3,
    FpAuthorized = 4,
    FpStarted = 5,
    FpFuelling = 6,
    FpSuspended = 7,
    FpCompleted = 8,
    FpLocked = 9,
    FpError = 10,
    FpEmergencyStop = 11,
    FpDisconnected = 12,
    FpOffline = 13
}

public static class DomsFpMainStateExtensions
{
    /// <summary>
    /// Maps a DOMS FP_MAIN_STATE value to the canonical <see cref="PumpState"/> enum
    /// shared across all FCC adapter implementations.
    /// </summary>
    public static PumpState ToCanonicalPumpState(this DomsFpMainState state) => state switch
    {
        DomsFpMainState.FpInoperative    => PumpState.Offline,
        DomsFpMainState.FpClosed         => PumpState.Offline,
        DomsFpMainState.FpIdle           => PumpState.Idle,
        DomsFpMainState.FpCalling        => PumpState.Calling,
        DomsFpMainState.FpAuthorized     => PumpState.Authorized,
        DomsFpMainState.FpStarted        => PumpState.Authorized,
        DomsFpMainState.FpFuelling       => PumpState.Dispensing,
        DomsFpMainState.FpSuspended      => PumpState.Paused,
        DomsFpMainState.FpCompleted      => PumpState.Completed,
        DomsFpMainState.FpLocked         => PumpState.Offline,
        DomsFpMainState.FpError          => PumpState.Error,
        DomsFpMainState.FpEmergencyStop  => PumpState.Error,
        DomsFpMainState.FpDisconnected   => PumpState.Offline,
        DomsFpMainState.FpOffline        => PumpState.Offline,
        _                                => PumpState.Unknown,
    };
}

File: src/edge-agent/app/src/main/kotlin/com/fccmiddleware/edge/adapter/doms/model/DomsFpMainState.kt
package com.fccmiddleware.edge.adapter.doms.model

import com.fccmiddleware.edge.adapter.common.PumpState

/**
 * DOMS FP_MAIN_STATE values from the DOMS protocol specification.
 * Maps raw integer state codes to the canonical [PumpState] enum.
 */
enum class DomsFpMainState(val code: Int) {
    FP_INOPERATIVE(0),
    FP_CLOSED(1),
    FP_IDLE(2),
    FP_CALLING(3),
    FP_AUTHORIZED(4),
    FP_STARTED(5),
    FP_FUELLING(6),
    FP_SUSPENDED(7),
    FP_COMPLETED(8),
    FP_LOCKED(9),
    FP_ERROR(10),
    FP_EMERGENCY_STOP(11),
    FP_DISCONNECTED(12),
    FP_OFFLINE(13);

    /** Convert this DOMS-specific pump state to the canonical [PumpState]. */
    fun toCanonicalPumpState(): PumpState = when (this) {
        FP_INOPERATIVE -> PumpState.OFFLINE
        FP_CLOSED -> PumpState.OFFLINE
        FP_IDLE -> PumpState.IDLE
        FP_CALLING -> PumpState.CALLING
        FP_AUTHORIZED -> PumpState.AUTHORIZED
        FP_STARTED -> PumpState.AUTHORIZED   // G-13 fix: started = authorized, not yet dispensing
        FP_FUELLING -> PumpState.DISPENSING
        FP_SUSPENDED -> PumpState.PAUSED
        FP_COMPLETED -> PumpState.COMPLETED
        FP_LOCKED -> PumpState.OFFLINE        // G-13 fix: locked = offline, must not accept transactions
        FP_ERROR -> PumpState.ERROR
        FP_EMERGENCY_STOP -> PumpState.ERROR
        FP_DISCONNECTED -> PumpState.OFFLINE
        FP_OFFLINE -> PumpState.OFFLINE
    }

    companion object {
        private val byCode = entries.associateBy { it.code }

        /** Look up a [DomsFpMainState] by its integer code. Returns null if unknown. */
        fun fromCode(code: Int): DomsFpMainState? = byCode[code]
    }
}
